File: CashuWallet/Core/WalletStore.swift
import Foundation

final class WalletStore {
    private let storage: StorageProtocol

    init(storage: StorageProtocol = UserDefaultsStorage()) {
        self.storage = storage
    }

    var activeMintURL: String? {
        get { value(forKey: StorageKeys.activeMintUrl) }
        set { setOptional(newValue, forKey: StorageKeys.activeMintUrl) }
    }

    func loadMints() -> [MintInfo] {
        value(forKey: StorageKeys.mints, legacyKeys: [StorageKeys.Legacy.mints]) ?? []
    }

    func saveMints(_ mints: [MintInfo]) {
        set(mints, forKey: StorageKeys.mints)
    }

    func loadPendingTokens() -> [PendingToken] {
        value(forKey: StorageKeys.pendingTokens, legacyKeys: [StorageKeys.Legacy.pendingTokens]) ?? []
    }

    func savePendingTokens(_ tokens: [PendingToken]) {
        set(tokens, forKey: StorageKeys.pendingTokens)
    }

    func loadPendingReceiveTokens() -> [PendingReceiveToken] {
        value(
            forKey: StorageKeys.pendingReceiveTokens,
            legacyKeys: [StorageKeys.Legacy.pendingReceiveTokens]
        ) ?? []
    }

    func savePendingReceiveTokens(_ tokens: [PendingReceiveToken]) {
        set(tokens, forKey: StorageKeys.pendingReceiveTokens)
    }

    func loadClaimedTokens() -> [ClaimedToken] {
        value(forKey: StorageKeys.claimedTokens, legacyKeys: [StorageKeys.Legacy.claimedTokens]) ?? []
    }

    func saveClaimedTokens(_ tokens: [ClaimedToken]) {
        set(tokens, forKey: StorageKeys.claimedTokens)
    }

    func loadSavedTokens() -> [String: String] {
        value(forKey: StorageKeys.savedTokens, legacyKeys: [StorageKeys.Legacy.savedTokens]) ?? [:]
    }

    func saveSavedTokens(_ tokens: [String: String]) {
        set(tokens, forKey: StorageKeys.savedTokens)
    }

    func loadPaymentPreimages() -> [String: String] {
        value(
            forKey: StorageKeys.paymentPreimages,
            legacyKeys: [StorageKeys.Legacy.paymentPreimages]
        ) ?? [:]
    }

    func savePaymentPreimages(_ preimages: [String: String]) {
        set(preimages, forKey: StorageKeys.paymentPreimages)
    }

    func loadMeltQuoteFees() -> [String: UInt64] {
        value(forKey: StorageKeys.meltQuoteFees) ?? [:]
    }

    func saveMeltQuoteFees(_ fees: [String: UInt64]) {
        set(fees, forKey: StorageKeys.meltQuoteFees)
    }

    func loadMintQuoteTimestamps() -> [String: TimeInterval] {
        value(
            forKey: StorageKeys.mintQuoteTimestamps,
            legacyKeys: [StorageKeys.Legacy.mintQuoteTimestamps]
        ) ?? [:]
    }

    func saveMintQuoteTimestamps(_ timestamps: [String: TimeInterval]) {
        set(timestamps, forKey: StorageKeys.mintQuoteTimestamps)
    }

    func loadProcessedNPCQuotes() -> [String] {
        value(forKey: StorageKeys.processedNPCQuotes) ?? []
    }

    func saveProcessedNPCQuotes(_ quoteIds: [String]) {
        set(quoteIds, forKey: StorageKeys.processedNPCQuotes)
    }

    func removeAllWalletData() {
        remove(keys: StorageKeys.walletDataKeys + StorageKeys.walletDataLegacyKeys)
        remove(keys: storage.keys(withPrefix: StorageKeys.walletDataPrefix))
    }

    private func value<T: Codable>(forKey key: String, legacyKeys: [String] = []) -> T? {
        if let value: T = try? storage.get(forKey: key) {
            return value
        }

        for legacyKey in legacyKeys {
            if let value: T = try? storage.get(forKey: legacyKey) {
                set(value, forKey: key)
                return value
            }
        }

        return nil
    }

    private func set<T: Codable>(_ value: T, forKey key: String) {
        do {
            try storage.set(value, forKey: key)
        } catch {
            AppLogger.wallet.error("Failed to save \(key): \(error)")
        }
    }

    private func setOptional<T: Codable>(_ value: T?, forKey key: String) {
        do {
            if let value {
                try storage.set(value, forKey: key)
            } else {
                try storage.remove(forKey: key)
            }
        } catch {
            AppLogger.wallet.error("Failed to update \(key): \(error)")
        }
    }

    private func remove(keys: [String]) {
        for key in Set(keys) {
            do {
                try storage.remove(forKey: key)
            } catch {
                AppLogger.wallet.error("Failed to remove \(key): \(error)")
            }
        }
    }
}

File: android/app/src/main/java/com/cashu/me/Core/WalletStore.kt
package com.cashu.me.Core

import android.content.Context
import kotlinx.serialization.KSerializer
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import com.cashu.me.Core.Protocols.StorageKeys
import com.cashu.me.Models.CashuRequest
import com.cashu.me.Models.ClaimedToken
import com.cashu.me.Models.MintInfo
import com.cashu.me.Models.PendingReceiveToken
import com.cashu.me.Models.PendingToken
import com.cashu.me.Models.WalletTransaction

class WalletStore(
    context: Context,
    storeName: String = "wallet_store",
) : CashuRequestPersistence {
    private val store = DataStorePreferenceStore(context.applicationContext, storeName)
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    var activeMintURL: String?
        get() = store.string(StorageKeys.walletActiveMintUrl)
        set(value) = store.putString(StorageKeys.walletActiveMintUrl, value)

    fun loadMints(): List<MintInfo> = loadList(StorageKeys.walletMints, MintInfo.serializer())
    fun saveMints(mints: List<MintInfo>) = saveList(StorageKeys.walletMints, MintInfo.serializer(), mints)

    fun loadPendingTokens(): List<PendingToken> = loadList(StorageKeys.walletPendingTokens, PendingToken.serializer())
    fun savePendingTokens(tokens: List<PendingToken>) = saveList(StorageKeys.walletPendingTokens, PendingToken.serializer(), tokens)

    fun loadPendingReceiveTokens(): List<PendingReceiveToken> = loadList(StorageKeys.walletPendingReceiveTokens, PendingReceiveToken.serializer())
    fun savePendingReceiveTokens(tokens: List<PendingReceiveToken>) =
        saveList(StorageKeys.walletPendingReceiveTokens, PendingReceiveToken.serializer(), tokens)

    fun loadClaimedTokens(): List<ClaimedToken> = loadList(StorageKeys.walletClaimedTokens, ClaimedToken.serializer())
    fun saveClaimedTokens(tokens: List<ClaimedToken>) = saveList(StorageKeys.walletClaimedTokens, ClaimedToken.serializer(), tokens)

    fun loadTransactions(): List<WalletTransaction> = loadList(StorageKeys.walletTransactions, WalletTransaction.serializer())
    fun saveTransactions(transactions: List<WalletTransaction>) =
        saveList(StorageKeys.walletTransactions, WalletTransaction.serializer(), transactions)

    fun loadPaymentPreimages(): Map<String, String> =
        loadMap(StorageKeys.walletPaymentPreimages, String.serializer())
    fun savePaymentPreimages(preimages: Map<String, String>) =
        saveMap(StorageKeys.walletPaymentPreimages, String.serializer(), preimages)

    fun loadMeltQuoteFees(): Map<String, Long> =
        loadMap(StorageKeys.walletMeltQuoteFees, Long.serializer())
    fun saveMeltQuoteFees(fees: Map<String, Long>) =
        saveMap(StorageKeys.walletMeltQuoteFees, Long.serializer(), fees)

    fun loadMintQuoteTimestamps(): Map<String, Long> =
        loadMap(StorageKeys.walletMintQuoteTimestamps, Long.serializer())
    fun saveMintQuoteTimestamps(timestamps: Map<String, Long>) =
        saveMap(StorageKeys.walletMintQuoteTimestamps, Long.serializer(), timestamps)

    fun loadProcessedNPCQuotes(): List<String> = loadList(StorageKeys.walletProcessedNPCQuotes, String.serializer())
    fun saveProcessedNPCQuotes(quotes: List<String>) =
        saveList(StorageKeys.walletProcessedNPCQuotes, String.serializer(), quotes)

    fun loadProcessedCashuRequests(): List<String> =
        loadList(StorageKeys.walletProcessedCashuRequests, String.serializer())
    fun saveProcessedCashuRequests(requestIds: List<String>) =
        saveList(StorageKeys.walletProcessedCashuRequests, String.serializer(), requestIds)

    override fun loadCashuRequests(): List<CashuRequest> =
        loadList(StorageKeys.cashuRequests, CashuRequest.serializer()).map { it.withLegacyPaymentFallback() }
    override fun saveCashuRequests(requests: List<CashuRequest>) =
        saveList(StorageKeys.cashuRequests, CashuRequest.serializer(), requests.map { it.withLegacyPaymentFallback() })

    override var currentCashuRequestId: String?
        get() = store.string(StorageKeys.cashuRequestsCurrentId)
        set(value) = store.putString(StorageKeys.cashuRequestsCurrentId, value)

    internal fun snapshotWalletScopedData(): PreferenceSnapshot {
        val prefixKeys = store.keys().filter {
            it.startsWith(StorageKeys.walletDataPrefix) || it.startsWith(StorageKeys.npcDataPrefix)
        }
        return store.snapshot(StorageKeys.walletBoundaryKeys + prefixKeys)
    }

    internal fun restoreWalletScopedData(snapshot: PreferenceSnapshot) {
        store.restore(snapshot)
    }

    fun removeAllWalletData() {
        store.removeKeys(StorageKeys.walletBoundaryKeys)
        store.removePrefix(listOf(StorageKeys.walletDataPrefix, StorageKeys.npcDataPrefix))
    }

    private fun <T> loadList(key: String, serializer: KSerializer<T>): List<T> {
        val raw = store.string(key) ?: return emptyList()
        return runCatching { json.decodeFromString(ListSerializer(serializer), raw) }.getOrDefault(emptyList())
    }

    private fun <T> saveList(key: String, serializer: KSerializer<T>, values: List<T>) {
        store.putString(key, json.encodeToString(ListSerializer(serializer), values))
    }

    private fun <T> loadMap(key: String, serializer: KSerializer<T>): Map<String, T> {
        val raw = store.string(key) ?: return emptyMap()
        return runCatching { json.decodeFromString(MapSerializer(String.serializer(), serializer), raw) }
            .getOrDefault(emptyMap())
    }

    private fun <T> saveMap(key: String, serializer: KSerializer<T>, values: Map<String, T>) {
        store.putString(key, json.encodeToString(MapSerializer(String.serializer(), serializer), values))
    }
}
